libkbfs/folder_branch_ops.go

```diff
@@ -3787,10 +3741,85 @@ func (fbo *folderBranchOps) syncAllLocked(
 			fbo.dirOps = nil
 		}
 	}()
-	for _, op := range fbo.dirOps {
-		md.AddOp(op.dirOp)
+
+	newBlocks := make(map[BlockPointer]bool)
+	fileBlocks := make(fileBlockMap)
+	parentsToAddChainsFor := make(map[BlockPointer]bool)
+	for _, dop := range fbo.dirOps {
+		// Copy the op before modifying it, in case there's an error
+		// and we have to retry with the original ops.  TODO: make a
+		// better way of copying a single op.
+		ops := make(opsList, 1)
+		err := kbfscodec.Update(
+			fbo.config.Codec(), &ops, opsList([]op{dop.dirOp}))
+		if err != nil {
+			return err
+		}
+		newOp := ops[0]
+
+		md.AddOp(newOp)
+
+		// Add "updates" for all the op updates, and make chains for
+		// the rest of the parent directories, so they're treated like
+		// updates during the prepping.
+		for _, n := range dop.nodes {
+			p := fbo.nodeCache.PathFromNode(n)
+			if _, ok := newOp.(*setAttrOp); ok {
+				// For a setattr, the node is the file, but that
+				// doesn't get updated, so use the current parent
+				// node.
+				p = *p.parentPath()
+			}
+
+			for i, pn := range p.path {
+				if i == len(p.path)-1 {
+					newOp.AddUpdate(pn.BlockPointer, pn.BlockPointer)
+				} else {
+					parentsToAddChainsFor[pn.BlockPointer] = true
+				}
+			}
+		}
+
 		var ref BlockRef
-		switch realOp := op.dirOp.(type) {
+		switch realOp := newOp.(type) {
+		case *createOp:
+			if realOp.Type == Sym {
+				continue
+			}
+
+			// New files and directories need explicitly
+			// pointer-updating, because the sync process will turn
+			// them into simple refs and will forget about the local,
+			// temporary ID.
+			newNode := dop.nodes[1]
+			newPath := fbo.nodeCache.PathFromNode(newNode)
+			newBlocks[newPath.tailPointer()] = true
+
+			if realOp.Type != Dir {
+				continue
+			}
+
+			dblock, ok := lbc[newPath.tailPointer()]
+			if !ok {
+				// New directories that aren't otherwise dirty need to
+				// be added to both the `lbc` and `resolvedPaths` so
+				// they are properly synced.
+				dblock, err = fbo.blocks.GetDirtyDir(
+					ctx, lState, md, newPath, blockWrite)
+				if err != nil {
+					return err
+				}
+				lbc[newPath.tailPointer()] = dblock
+				resolvedPaths[newPath.tailPointer()] = newPath
+			}
+
+			if len(dblock.Children) > 0 {
+				continue
+			}
+
+			// If the directory is empty, we need to explicitly clean
```

Comment:
If the directory *isn't* empty, it's guaranteed that we'll process other createOps for that directory, and we'll set it to be cleaned up at that point.